cpp/src/strings/replace/multi.cu

```diff
@@ -0,0 +1,488 @@
+  }
+  replace_multi_parallel_fn fn;
+  size_type chars_bytes;
+};
+
+struct copy_if_fn {
+  __device__ bool operator()(target_pair pos) { return pos.second >= 0; }
+};
+
+/**
+ * @brief Function logic for the replace_multi API.
+ *
+ * Performs the multi-replace operation with a thread per string.
+ * This performs best on smaller strings. @see AVG_CHAR_BYTES_THRESHOLD
+ */
+struct replace_multi_fn {
+  column_device_view const d_strings;
+  column_device_view const d_targets;
+  column_device_view const d_repls;
+  int32_t* d_offsets{};
+  char* d_chars{};
+
+  __device__ void operator()(size_type idx)
+  {
+    if (d_strings.is_null(idx)) {
+      if (!d_chars) d_offsets[idx] = 0;
+      return;
+    }
+    auto const d_str   = d_strings.element<string_view>(idx);
+    char const* in_ptr = d_str.data();
+
+    size_type bytes = d_str.size_bytes();
+    size_type spos  = 0;
+    size_type lpos  = 0;
+    char* out_ptr   = d_chars ? d_chars + d_offsets[idx] : nullptr;
+
+    // check each character against each target
+    while (spos < d_str.size_bytes()) {
+      for (int tgt_idx = 0; tgt_idx < d_targets.size(); ++tgt_idx) {
+        auto const d_tgt = d_targets.element<string_view>(tgt_idx);
+        if ((d_tgt.size_bytes() <= (d_str.size_bytes() - spos)) &&    // check fit
+            (d_tgt.compare(in_ptr + spos, d_tgt.size_bytes()) == 0))  // and match
+        {
+          auto const d_repl = (d_repls.size() == 1) ? d_repls.element<string_view>(0)
+                                                    : d_repls.element<string_view>(tgt_idx);
+          bytes += d_repl.size_bytes() - d_tgt.size_bytes();
+          if (out_ptr) {
+            out_ptr = copy_and_increment(out_ptr, in_ptr + lpos, spos - lpos);
+            out_ptr = copy_string(out_ptr, d_repl);
+            lpos    = spos + d_tgt.size_bytes();
+          }
+          spos += d_tgt.size_bytes() - 1;
+          break;
+        }
+      }
+      ++spos;
+    }
+    if (out_ptr)  // copy remainder
+      memcpy(out_ptr, in_ptr + lpos, d_str.size_bytes() - lpos);
+    else
+      d_offsets[idx] = bytes;
+  }
+};
+
+}  // namespace
+
+std::unique_ptr<column> replace(strings_column_view const& input,
+                                strings_column_view const& targets,
+                                strings_column_view const& repls,
+                                rmm::cuda_stream_view stream,
+                                rmm::mr::device_memory_resource* mr)
+{
+  if (input.is_empty()) return make_empty_column(type_id::STRING);
+  CUDF_EXPECTS(((targets.size() > 0) && (targets.null_count() == 0)),
+               "Parameters targets must not be empty and must not have nulls");
+  CUDF_EXPECTS(((repls.size() > 0) && (repls.null_count() == 0)),
+               "Parameters repls must not be empty and must not have nulls");
+  if (repls.size() > 1)
+    CUDF_EXPECTS(repls.size() == targets.size(), "Sizes for targets and repls must match");
```

Comment:
```suggestion
  CUDF_EXPECTS(repls.size() == 1 || repls.size() == targets.size(), "Must provide a single replacement string or sizes for targets and replacements must match");
```

Reply:
I see this was copied code now. Unsure if you want to take this or not, but I'm leaving it so you can decide.